File: 385.py
# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """
#class NestedInteger(object):
#    def __init__(self, value=None):
#        """
#        If value is not specified, initializes an empty list.
#        Otherwise initializes a single integer equal to value.
#        """
#
#    def isInteger(self):
#        """
#        @return True if this NestedInteger holds a single integer, rather than a nested list.
#        :rtype bool
#        """
#
#    def add(self, elem):
#        """
#        Set this NestedInteger to hold a nested list and adds a nested integer elem to it.
#        :rtype void
#        """
#
#    def setInteger(self, value):
#        """
#        Set this NestedInteger to hold a single integer equal to value.
#        :rtype void
#        """
#
#    def getInteger(self):
#        """
#        @return the single integer that this NestedInteger holds, if it holds a single integer
#        Return None if this NestedInteger holds a nested list
#        :rtype int
#        """
#
#    def getList(self):
#        """
#        @return the nested list that this NestedInteger holds, if it holds a nested list
#        Return None if this NestedInteger holds a single integer
#        :rtype List[NestedInteger]
#        """

class Solution(object):
    def deserialize(self, s):
        """
        :type s: str
        :rtype: NestedInteger
        """
        if len(s) == 0:
          return

        if s[0] != '[':
          return NestedInteger(int(s))

        q = []
        number = ''

        for i in range(len(s)):
            if s[i] == '[':
                # q.append(s[i])
                q.append(NestedInteger())
            elif s[i] == ']':
                # temp = []
                # while q[-1] != '[':
                #     temp.insert(0, q.pop())
                # q.pop()
                # q.append(temp)
                temp = q.pop()
                if q:
                    q[-1].add(temp)
                else:
                    return temp

            elif s[i] == ',':
                if number != '':
                    q[-1].add(NestedInteger(int(number)))
                    number = ''
            # if s[i] == '-' or '0' <= s[i] <= '9':
            #     number += s[i]
            # if s[i+1] == ',' or s[i+1] == '[' or s[i+1] == ']':
            #     q.append(int(number))
            #     number = ''
            if s[i] == '-' or '0' <= s[i] <= '9':
                number += s[i]
            if number != '' and (s[i+1] == ',' or s[i+1] == ']'):
                q[-1].add(NestedInteger(int(number)))
                number = ''
           

File: Рудович/385.cpp
#include <vector>
#include <string>
#include <stack>

using namespace std;

class NestedInteger {
  public:
    // Constructor initializes an empty nested list.
    NestedInteger();

    // Constructor initializes a single integer.
    NestedInteger(int value);

    // Return true if this NestedInteger holds a single integer, rather than a nested list.
    bool isInteger() const;

    // Return the single integer that this NestedInteger holds, if it holds a single integer
    // The result is undefined if this NestedInteger holds a nested list
    int getInteger() const;

    // Set this NestedInteger to hold a single integer.
    void setInteger(int value);

    // Set this NestedInteger to hold a nested list and adds a nested integer to it.
    void add(const NestedInteger &ni);

    // Return the nested list that this NestedInteger holds, if it holds a nested list
    // The result is undefined if this NestedInteger holds a single integer
    const vector<NestedInteger> &getList() const;
};
class Solution {
public:
    NestedInteger deserialize(string s) 
    {
        if (s[0] == '[') return deserialize(s, 1).first;
        return deserialize(s, 0).first;
    }
    pair<NestedInteger, int> deserialize(string s, int i) 
    {
        NestedInteger ans;
        int num = 0;
        int sign = 0;
        for (i; i < s.size(); ++i)
        {
            if (s[i] == '-') sign = 2;
            else if (s[i] >= '0' && s[i] <= '9')
            {
                if (sign == 0) sign = 1;
                if (sign == 1) num = num * 10 + (s[i] - '0');
                else num = num * 10 - (s[i] - '0');
            }
            else if (s[i] == ',') 
            {
                if (sign != 0) ans.add(NestedInteger(num));
                sign = 0;
                num = 0;
            }
            else if (s[i] == '[')
            {
                pair<NestedInteger, int> res = deserialize(s, i + 1);
                ans.add(res.first);
                i = res.second;
                sign = 0;
                num = 0;
            }
            else
            {
                if (sign != 0) ans.add(NestedInteger(num));
                sign = 0;
                num = 0;
                return {ans, i};
            }
        }
        if (sign != 0)
        {
            if (ans.isInteger() || ans.getList().size() > 0) ans.add(NestedInteger(num));
            else ans.setInteger(num);
        }
        sign = 0;
        num = 0;
        return {ans, s.size()};
    }
};